### PR TITLE
update createSandboxContainer to parse hugepages limits from CRI message

### PR DESCRIPTION
**Pull request by bg-chun**
### The purpose of this PR
- updating vendor to refer the last CRI Update which includes hugepages update
- updating CRI-O to parse newly added hugepages fields from CRI message.

### History
- In Kubernetes, hugepages are requested as the container level resources, but limit is set on only pod level cgroup due to the lack of CRI support.
- But now, CRI has been updated to have hugepages field so that kubernetes be able to set hugepages limit on container level cgroup.
- The last step we have to do is updating container runtime which handles container level cgroup to parse new hugepage field from CRI message and set limit on container level cgroup.

### The Detail of CRI Update
- Now, LinuxContainerResources CRI message, which has linux resource fileds that equals to linux cgroup subsystems , has hugepage field(`hugepage_limits`), it is able to express list of hugepage configuration.
- New hugepage fields are written based on existing hugepage fields of [OCI runtime spec](https://github.com/opencontainers/runtime-spec/blob/master/config-linux.md#huge-page-limits) so that is is easy to transfrom hugepages fields and values from CRI message to OCI Runtime configuration.
- Each hugepage configuration(`HugepageLimit`) expesss hugepage limit per page size.
- The configuration [page_size:  "2MB", limit: 209715200] means set 2GB limits for 2MB hugepage size on the container level cgroup sandbox.
- More details are available on this [slide](https://docs.google.com/presentation/d/1J_6RGeNU1f21TchLrE672SfFLcbhPUev356NLGYxoUU/edit#slide=id.g7a7421ffe7_0_157) and [doc](https://docs.google.com/document/d/1Xnb3ioEH_JmWLdFJ-i6Lp7jFc_fDuzzAa2oEXJglBbM/edit#heading=h.va9fkr521iz7).

```
message LinuxContainerResources {
    ...
    string cpuset_mems = 7;
    // List of HugepageLimits to limit the HugeTLB usage of container per page size. Default: nil (not specified).
    repeated HugepageLimit hugepage_limits = 8;
}
// HugepageLimit corresponds to the file`hugetlb.<hugepagesize>.limit_in_byte` in container level cgroup.
// For example, `PageSize=1GB`, `Limit=1073741824` means setting `1073741824` bytes to hugetlb.1GB.limit_in_bytes.
message HugepageLimit {
    // The value of PageSize has the format <size><unit-prefix>B (2MB, 1GB),
    // and must match the <hugepagesize> of the corresponding control file found in `hugetlb.<hugepagesize>.limit_in_bytes`.
    // The values of <unit-prefix> are intended to be parsed using base 1024("1KB" = 1024, "1MB" = 1048576, etc).
    string page_size = 1;
    // limit in bytes of hugepagesize HugeTLB usage.
    uint64 limit = 2;
}
```
- KEP updates : https://github.com/kubernetes/enhancements/pull/1199
- CRI Updates : https://github.com/kubernetes/kubernetes/pull/83614

### CODE REVIEW

**Comment by saschagrunert** [comment]
/ok-to-test

Thanks, @bg-chun :pray:

**Comment by mrunalp** [comment]
@bg-chun We need to update cri api in go.mod for pulling in the proto changes. Thanks!

**Comment by bg-chun** [comment]
I will update pr after CRI pr merged on kubernetes.

File: server/container_create_linux.go [hunk 1]
```diff
@@ -491,6 +491,13 @@ func (s *Server) createSandboxContainer(ctx context.Context, containerID, contai
 			specgen.SetProcessOOMScoreAdj(int(resources.GetOomScoreAdj()))
 			specgen.SetLinuxResourcesCPUCpus(resources.GetCpusetCpus())
 			specgen.SetLinuxResourcesCPUMems(resources.GetCpusetMems())
+
+			hugepageLimits := resources.GetHugepageLimits()
+			if len(hugepageLimits) != 0 {
```
Comment (giuseppe) [inline server/container_create_linux.go]: is the `if len(hugepageLimits) != 0` check redundant?

Comment (rhatdan) [reply server/container_create_linux.go]: Yes remove this the range command below will handle this situation.

**Comment by bg-chun** [comment]
cross posting from cri-o slack channel to github.
https://kubernetes.slack.com/archives/CAZH62UR1/p1576735156005800

I'm trying to update cri-api vendor to new release tag.
(I've used vendor.conf, using go module is the fist time.)
So i tried below, my expectation was below command will update vendored cri-api(vendor/k8s.io/cri-api) to new alpha tag for 1.18.
go get -u k8s.io/cri-api@kubernetes-1.18.0-alpha.1
But what i got is not equal to my expectation.
I think go.mod is updated well, but vendored code(vendor/k8s.io/cri-api) is not updated.
What i have to do to update vendored cri-api?
```
bgchun@bgchun-labtop:~/go/src/github.com/cri-o$ git diff
+ git diff
diff --git a/go.mod b/go.mod
index 64b6e4dde..fcb675960 100644
--- a/go.mod
+++ b/go.mod
@@ -58,7 +58,7 @@ require (
        k8s.io/api v0.0.0
        k8s.io/apimachinery v0.0.0
        k8s.io/client-go v0.0.0
-       k8s.io/cri-api v0.0.0
+       k8s.io/cri-api v0.18.0-alpha.1
        k8s.io/kubernetes v1.13.0
        k8s.io/utils v0.0.0-20191114200735-6ca3b61696b6
 )
diff --git a/vendor/modules.txt b/vendor/modules.txt
index 9e355dca8..20c14a0e2 100644
--- a/vendor/modules.txt
+++ b/vendor/modules.txt
@@ -1002,7 +1002,7 @@ k8s.io/component-base/featuregate
 k8s.io/component-base/metrics
 k8s.io/component-base/metrics/legacyregistry
 k8s.io/component-base/version
-# k8s.io/cri-api v0.0.0 => k8s.io/kubernetes/staging/src/k8s.io/cri-api v0.0.0-20191206012503-70132b0f130a
+# k8s.io/cri-api v
```

File: server/container_create_linux.go (hunk 1 above)
Comment (bg-chun) [reply server/container_create_linux.go]: done

**Comment by bg-chun** [comment]
The description of PR has been updated :)

**Comment by mrunalp** [comment]
Looks fine to me. @saschagrunert you want to check the dependency update here?

**Comment by bg-chun** [comment]
@mrunalp
I think he checked it before. 
https://kubernetes.slack.com/archives/CAZH62UR1/p1576735156005800
BTW, thank you for your effort.

**Pull request merged**